README.md
# react-donut-chart

---

An extendable SVG-based donut chart React component.

![Example](https://github.com/vonbearshark/react-donut-chart/raw/master/example.gif)

---

## Installation

`npm install react-donut-chart --save`

---

## Usage

```js
import DonutChart from 'react-donut-chart';
//things I would never do:
<DonutChart
    data={[{
        label: 'Give you up',
        value: 25
    },
    {
        label: '',
        value: 75,
        isEmpty: true
    }]} />
```

View the demo [online](http://vonbearshark.github.io/react-donut-chart), or run it locally:

`npm run build`
`npm run start`

then visit [localhost:3000](http://localhost:3000).

---

## Props

Name | Type | isRequired | Default | Description
--- | --- | --- | --- | ---
data | <code>arrayOf(shape({<br>&nbsp;&nbsp;value: number.isRequired,<br>&nbsp;&nbsp;label: string,<br>&nbsp;&nbsp;className: string,<br>&nbsp;&nbsp;isEmpty: boolean<br>}))</code> | true | <code>[{<br>&nbsp;&nbsp;value: 100,<br>&nbsp;&nbsp;label: '',<br>&nbsp;&nbsp;isEmpty:true<br>}]</code> | The chart data
className | `string` | false | `donutchart` | This is the `className` given to the top-level `svg` element. All subclasses are prefixed from this name: <ul><li><code>${className}-arcs</code> accesses the entire graph area</li><ul><li><code>${className}-arcs-paths</code> accesses the individual arc paths</li></ul><li><code>${className}-innertext</code> accesses all of the text within the inner donut area</li><ul><li><code>${className}-innertext-label</code> accesses the label within this area</li><li><code>${className}-innertext-value</code> accesses the value within this area</li></ul><li><code>${className}-legend</code> accesses the legend component</li><ul><li><code>${className}-legend-rect</code> accesses the legend rectangle items</li><li><code>${className}-legend-label</code> accesses the labels of the legend items</li></ul></ul> In addition the `selected` class is given to selected items, the `toggled` class to all toggled items, and the `isEmpty` class to all `isEmpty` items. All style (and animations) can be manipulated from the CSS
height | `number` | false | `500` | Height of the entire component
width | `number` | false | `750` | Width of the entire component. If no legend is specified, then the chart takes up the entire width. If a legend is toggled, then the chart takes up 2/3 of the width, and the legend takes up 1/3
colors | `arrayOf(string)` | false | `['#f44336', '#e91e63', '#9c27b0', '#673ab7', '#3f51b5', '#2196f3', '#03a9f4', '#00bcd4', '#009688', '#4caf50', '#8bc34a', '#cddc39', '#ffeb3b', '#ffc107', '#ff9800', '#ff5722', '#795548', '#607d8b' ]` | An array of colors (could be hex strings or named colors) for the data items. Defaults to an array of Google colors
emptyColor | `string` | false | `'#e0e0e0'` | A color for empty data items, defaults to gray
stokeColor | `string` | false | `'#212121'` | A color for the stroke around the items in the graph and legend, defaults to black
colorFunction | `func` | false | `(colors, index) => colors[(index % colors.length)]` | The default cycles through the array of colors and loops for excess
innerRadius | `number` | false | `0.70` | The inner donut radius
outerRadius | `number` | false | `0.90` | The outer donut radius
selectedOffset | `number` | false | `0.03` | The `outerRadius` offset when an item is selected
emptyOffset | `number` | false | `0.08` | The `innerRadius` and `outerRadius` offset on `isEmpty` items
toggledOffset | `number` | false | `0.04` | The `innerRadius` and `outerRadius` offset on toggle-clicked items
startAngle | `number` | false | `0` | The drawing start angle
formatValues | `func` | false | ```(values, total) => `${(values / total * 100).toFixed(2)}%` ``` | Custom format for values displayed in the donut chart's inner text area. By default formats as percentages rounded to two decimal places.
onMouseEnter | `func` | false | `(item) => item` | Callback that fires when an item is hovered
onClick | `func` | false | `onClick: (item, toggled) => (toggled ? item : null)` | Callback that fires when an item is toggle-clicked
legend | `bool` | false | `true` | Determines whether or not to create a legend
clickToggle | `bool` | false | `true` | Determines whether or not to toggle-freeze the graph on the arc that has been clicked

---

## License

MIT


CHANGELOG.md
# Changelog

## 1.1.7
- Updated dependencies for bug fixes
- Updated linting

##

...

## 1.0.3
* Bug fixes when no data present and when 100%.

## 1.0.2:
* No legend offset by default
* Hacking for condition of 100%.

## 1.0.1:
* Adding lib dist build process to fix importing/

## 1.0.0:
* Initial release with demo


lib/src/DonutChart.js
/**
 * @fileOverview Main Donut Chart entry point.
 * Orchestrates all rendering for the chart, passing props to
 * Arcs, DonutInnerText, and Legend children components.
 * @name DonutChart.js
 * @author JJ Naughton
 * @license MIT
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Arcs from './Arcs';
import DonutInnerText from './DonutInnerText';
import Legend from './Legend';


/**
 * Calculates the total of all the items' value,
 * using a reduce function over the items.
 * @param {} items - An array of items, each with values, at least.
 */
function calculateTotal(items) {
    return items.reduce((sum, currItem) => sum + currItem.value, 0);
}

/**
 * @extends {Component}
 */
export default class DonutChart extends Component {
    /**
    * Represents a DonutChart.
    * @constructor
    * @param {} props - Object of propTypes with defaultProps as fallback.
    */
    constructor(props) {
        super(props);

        const { data } = props;

        this.state = {
            selected: data[0],
            toggleSelect: false
        };
    }

    componentWillReceiveProps(newProps) {
        const { data } = newProps;

        // if new data, reset
        if (data && (JSON.stringify(data)
                    !== JSON.stringify(this.props.data))) {
            this.setState({
                selected: data[0],
                toggleSelect: false
            });
        }
    }

    /**
     * Handles the clicking of an ArcPath region.
     * Toggles the selected state,
     * effectively freezing all handlers on other, unselected items.
     * Calls the onClick handler,
     * passed by the user as a prop,
     * providing the item clicked,
     * as well as whether it's becoming selected or unselected.
     * @param {} item - The item object selected, with a label and a value, at least.
     */
    handleClick(item) {
        if (this.state.selected.label === item.label) {
            const toggle = this.props.clickToggle ? !this.state.toggleSelect : false;
            this.setState({
                toggleSelect: toggle,
                selected: item
            });
            this.props.onClick(item, toggle);
        }
    }

    /**
     * Handles the mouseenter event over an ArcPath region.
     * Sets the clicked item as selected in the state object.
     * @param {} item - The item object selected, with a label and a value, at least.
     */
    handleMouseEnter(item) {
        if (!this.state.toggleSelect) {
            this.setState({
                selected: item
            });
            this.props.onMouseEnter(item);
        }
    }

    /* React render function */
    render() {
        const {
            startAngle, width, height, formatValues, className, data, legend,
            emptyColor, strokeColor, colors, colorFunction, innerRadius,
            outerRadius, emptyOffset, selectedOffset, toggledOffset
        } = this.props;

        const arcsClassName = `${className}-arcs`;
        const innerTextClassName = `${className}-innertext`;
        const legendClassName = `${className}-legend`;

        const checkData = data.length ? data : [{
            label: '',
            value: 100,
            isEmpty: true
        }];

        const total = calculateTotal(checkData);
        const twoThirds = 2 / 3;
        const graphWidth = legend ? (width * twoThirds) : width;
        const legendWidth = width - graphWidth;

        return <svg className={ className }
            width={ width }
            height={ height }
            viewBox={`0 0 ${width} ${height}`}>
            <Arcs
                className={ arcsClassName }
                colors={ colors }
                data={ checkData }
                width={ graphWidth }
                emptyColor={ emptyColor }
                strokeColor={ strokeColor }
                colorFunction={ colorFunction }
                onMouseEnter={ this.handleMouseEnter.bind(this) }
                onClick={ this.handleClick.bind(this) }
                selected={ this.state.selected }
                startAngle = { startAngle }
                toggleSelect={ this.state.toggleSelect }
                innerRadius={ innerRadius }
                outerRadius={ outerRadius }
                selectedOffset={ selectedOffset }
                toggledOffset={ toggledOffset }
                emptyOffset={ emptyOffset }
                total={ total } />
            <DonutInnerText
                item={ this.state.selected }
                width={ graphWidth }
                formatValues={ formatValues }
                total={ total }
                className= { innerTextClassName } />
            { legend ?
                <Legend
                    data={ checkData }
                    totalWidth={ width }
                    width={ legendWidth }
                    colors={ colors }
                    emptyColor={ emptyColor }
                    strokeColor={ strokeColor }
                    colorFunction={ colorFunction }
                    onMouseEnter={ this.handleMouseEnter.bind(this) }
                    onClick={ this.handleClick.bind(this) }
                    selected={ this.state.selected }
                    toggleSelect={ this.state.toggleSelect}
                    className={ legendClassName } />
                : null }
        </svg>;
    }
}

DonutChart.propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({
        value: PropTypes.number.isRequired,
        label: PropTypes.string.isRequired,
        className: PropTypes.string,
        isEmpty: PropTypes.boolean
    })).isRequired,
    className: PropTypes.string,
    height: PropTypes.number,
    width: PropTypes.number,
    colors: PropTypes.arrayOf(PropTypes.string),
    emptyColor: PropTypes.string,
    stokeColor: PropTypes.string,
    startAngle: PropTypes.number,
    colorFunction: PropTypes.func,
    innerRadius: PropTypes.number,
    outerRadius: PropTypes.number,
    selectedOffset: PropTypes.number,
    emptyOffset: PropTypes.number,
    toggledOffset: PropTypes.number,
    formatValues: PropTypes.func,
    onMouseEnter: PropTypes.func,
    onClick: PropTypes.func,
    legend: PropTypes.bool,
    clickToggle: PropTypes.bool
};

DonutChart.defaultProps = {
    data: [{
        label: '',
        value: 100,
        isEmpty: true
    }],
    className: 'donutchart',
    height: 500,
    width: 750,
    colors: ['#f44336', '#e91e63', '#9c27b0', '#673ab7',
        '#3f51b5', '#2196f3', '#03a9f4', '#00bcd4',
        '#009688', '#4caf50', '#8bc34a', '#cddc39',
        '#ffeb3b', '#ffc107', '#ff9800', '#ff5722',
        '#795548', '#607d8b'],
    emptyColor: '#e0e0e0',
    strokeColor: '#212121',
    startAngle: 0,
    colorFunction: (colors, index) => colors[(index % colors.length)],
    innerRadius: 0.70,
    outerRadius: 0.90,
    selectedOffset: 0.03,
    emptyOffset: 0.08,
    toggledOffset: 0.04,
    formatValues: (value, total) => (
        Number.isNaN(value / total) ? '--'
            : `${((value / total) * 100).toFixed(2)}%`
    ),
    onMouseEnter: item => item,
    onClick: (item, toggled) => (toggled ? item : null),
    legend: true,
    clickToggle: true
};
